sdk/python/kfp/dsl/_pipeline_volume.py

```diff
@@ -0,0 +1,147 @@
+    to be mounted by a ContainerOp or its inherited type.
+
+    A PipelineVolume object can be used as an extention of the pipeline
+    function's filesystem. It may then be passed between ContainerOps,
+    exposing dependencies.
+    """
+    def __init__(self,
+                 pvc: str = None,
+                 volume: V1Volume = None,
+                 **kwargs):
+        """Create a new instance of PipelineVolume.
+
+        Args:
+            pvc: The name of an existing PVC
+            volume: Create a deep copy out of a V1Volume or PipelineVolume
+                with no deps
+        Raises:
+            ValueError: if pvc is not None and name is None
+                        if volume is not None and kwargs is not None
+                        if pvc is not None and kwargs.pop("name") is not None
+        """
+        if pvc and "name" not in kwargs:
+            raise ValueError("Please provide name.")
+        elif volume and kwargs:
+            raise ValueError("You can't pass a volume along with other "
+                             "kwargs.")
+
+        init_volume = {}
+        if volume:
+            init_volume = {attr: getattr(volume, attr)
+                           for attr in self.attribute_map.keys()}
+        else:
+            init_volume = {"name": kwargs.pop("name")
+                           if "name" in kwargs else None}
+            if pvc and kwargs:
+                raise ValueError("You can only pass 'name' along with 'pvc'.")
+            elif pvc and not kwargs:
+                pvc_volume_source = V1PersistentVolumeClaimVolumeSource(
+                    claim_name=pvc
+                )
+                init_volume["persistent_volume_claim"] = pvc_volume_source
+        super().__init__(**init_volume, **kwargs)
+        self.deps = []
+
+    def after(self, *ops):
+        """Creates a duplicate of self with the required dependecies excluding
+        the redundant dependenices.
+        Args:
+            *ops: Pipeline operators to add as dependencies
+        """
+        def implies(newdep, olddep):
+            if newdep.name == olddep:
+                return True
+            for parentdep_name in newdep.deps:
+                if parentdep_name == olddep:
+                    return True
+                else:
+                    parentdep = _pipeline.Pipeline.get_default_pipeline(
+                    ).ops[parentdep_name]
+                    if parentdep:
+                        if implies(parentdep, olddep):
+                            return True
+            return False
+
+        ret = self.__class__(volume=self)
+        ret.deps = [op.name for op in ops]
+
+        for olddep in self.deps:
+            implied = False
+            for newdep in ops:
+                implied = implies(newdep, olddep)
+                if implied:
+                    break
+            if not implied:
+                ret.deps.append(olddep)
+
+        return ret
+
+    def snapshot(self,
```

Comment:
This method seems unnecessary. You can move all the logic here to VolumeSnapshotOp. 
So instead of volume.snapshot(), it can be VolumeSnapshotOp(volume)

Reply:
ACK. You prefer there is only a single way for creating a snapshot from a volume, using an explicit `VolumeSnapshotOp` template. We will push a small change to remove this.

Reply:
I removed this method, the changes are force-pushed.